pkg/dmaasbackup/periodic_backup_test.go

```diff
@@ -0,0 +1,602 @@
+
+				if test.scheduleCreateError {
+					// inject error
+					return true, nil, &apierrors.StatusError{
+						ErrStatus: metav1.Status{Reason: metav1.StatusReasonInvalid},
+					}
+				}
+				// set creationTimestamp
+				setCreationTime := func(obj *velerov1api.Schedule) error {
+					var ntime time.Time
+
+					nslice := strings.Split(obj.Name, "-")
+					if len(nslice) != 2 {
+						return errors.New("invalid name")
+					}
+					ntime, err = time.Parse("20060102150405", nslice[1])
+					if err != nil {
+						return errors.New("invalid name format")
+					}
+					obj.CreationTimestamp.Time = ntime
+					return nil
+				}
+
+				err = setCreationTime(obj)
+				if err != nil {
+					return false, obj, err
+				}
+
+				_, exists := scheduleInventory[obj.Name]
+				if exists {
+					return true, obj, &apierrors.StatusError{
+						ErrStatus: metav1.Status{Reason: metav1.StatusReasonAlreadyExists},
+					}
+				}
+
+				// add schedule to scheduleInventory
+				scheduleInventory[obj.Name] = true
+
+				return true, obj, nil
+			})
+
+			if len(test.existingSchedules) != 0 {
+				for _, schedule := range test.existingSchedules {
+					pinnedSchedule := schedule
+					_, err = veleroclient.VeleroV1().Schedules("ns").Create(&pinnedSchedule)
+					require.NoError(t, err, "failed to setup existing schedule err=%v", err)
+				}
+			}
+
+			veleroclient.PrependReactor("delete", "schedules", func(action core.Action) (bool, runtime.Object, error) {
+				scheduleName := action.(core.DeleteAction).GetName()
+
+				if test.scheduleDeleteError {
+					// inject error
+					return true, nil, &apierrors.StatusError{
+						ErrStatus: metav1.Status{Reason: metav1.StatusReasonInvalid},
+					}
+				}
+				_, exists := scheduleInventory[scheduleName]
+				if exists {
+					delete(scheduleInventory, scheduleName)
+				} else {
+					return true, nil, &apierrors.StatusError{
+						ErrStatus: metav1.Status{Reason: metav1.StatusReasonNotFound},
+					}
+				}
+
+				return true, nil, nil
+			})
+
+			backupper := NewDMaaSBackupper(
+				"ns",
+				client,
+				veleroclient.VeleroV1(),
+				veleroFakeInformer.Velero().V1(),
+				apimachineryclock.NewFakeClock(testTime),
+			)
+
+			sort.Sort(sort.Reverse(ScheduleByCreationTimestamp(test.backup.Status.VeleroSchedules)))
+			bkpper, ok := backupper.(*dmaasBackup)
+			assert.True(t, ok, "failed to parse dmaasbackupper")
+
+			bkpper.logger = logrus.New()
+
+			err = bkpper.processPeriodicConfigSchedule(test.backup)
+
+			// verify error
+			if test.shouldErrored {
+				require.NotNil(t, err, "execute return nil error")
+			} else {
+				require.NoError(t, err, "execute return error err=%v", err)
+			}
+
+			// verify schedule list
+			matchError := compareVeleroSchedule(test.backup.Status.VeleroSchedules, test.expectedVeleroSchedule)
+			assert.NoError(t, matchError, "veleroschedule not matched")
+
+			// verify shouldRequeue
+			assert.Equal(t, test.shouldRequeue, bkpper.shouldRequeue, "Requeue validation failed")
```

Comment:
any cleanup of PrependReactor should be done inside the loop at the end so that it won't impact for next iteration?

Reply:
not required since on each iteration we are creating new clientset.